include/sys/zfs_ioctl.h

```diff
@@ -386,20 +387,27 @@ typedef struct dmu_replay_record {
 
 /* diff record range types */
 typedef enum diff_type {
-	DDR_NONE = 0x1,
-	DDR_INUSE = 0x2,
-	DDR_FREE = 0x4
+	DDR_NONE	= 0x0,
+	DDR_IN_TO	= 0x1,	/* dnode in to snap only */
+	DDR_IN_FROM	= 0x2,	/* dnode in from snap only */
+	DDR_IN_BOTH	= (DDR_IN_TO | DDR_IN_FROM),
 } diff_type_t;
 
 /*
  * The diff reports back ranges of free or in-use objects.
  */
 typedef struct dmu_diff_record {
-	uint64_t ddr_type;
-	uint64_t ddr_first;
-	uint64_t ddr_last;
+	diff_type_t	ddr_type;
+	uint64_t	ddr_obj;
+	int		ddr_err[2];
+	zfs_diff_stat_t	ddr_zds[2];	/* space for 1 or two stats as needed */
```

Comment:
"1" or.. "two"? also, isn't this fundamentally what a two-element array, uh, is?

Reply:
I presume they were trying to communicate that it's not always expected for both to be populated.